client/blb/client.go

```diff
@@ -1122,39 +1141,127 @@ func (cli *Client) readOneTractReplicated(
+	curAddr string,
+	host string,
+	otherHosts []string,
+	reqID string,
+	result *tractResult,
+	tract *core.TractInfo,
+	thisB []byte,
+	thisOffset int64) (skipHost bool, badHost string, err core.Error) {
+
+	var read int
+	var badVersionHost string
+	read, err = cli.tractservers.ReadInto(ctx, host, otherHosts, reqID, tract.Tract, tract.Version, thisB, thisOffset)
+	if err == core.ErrVersionMismatch {
+		badVersionHost = host
+	}
+	if err != core.NoError && err != core.ErrEOF {
+		log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
+		// If we failed to read from a TS, report that to the curator. Defer so we can examine
+		// result.err at the end, to see if we succeeded on another or not.
+		defer func(host string, err core.Error) {
+			couldRecover := result.err == core.NoError || result.err == core.ErrEOF
+			go cli.curators.ReportBadTS(context.Background(), curAddr, tract.Tract, host, "read", err, couldRecover)
+		}(host, err)
+		return true, badVersionHost, err // try another host
+
+	}
+	log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
+	// In case of short read, i.e., read < len(thisB), we should
+	// pad the result with 0's. This doesn't need to be done
+	// explicitly if 'thisB' given to us starts to be empty.
+	// However, just in case it's not, we will ensure that by
+	// overwritting untouched bytes with 0's.
+	for i := read; i < len(thisB); i++ {
+		thisB[i] = 0
+	}
+	*result = tractResult{len(thisB), read, err, badVersionHost}
+	return false, badVersionHost, err
+}
+
+func (cli *Client) readOneTractWithBackupReplicated(
+	ctx context.Context,
+	curAddr string,
+	result *tractResult,
+	tract *core.TractInfo,
+	thisB []byte,
+	thisOffset int64) {
+
+	var badVersionHost string
+
+	order := rand.Perm(len(tract.Hosts))
+
+	// Generate a request ID to use for cancelling alternate requests.
+	reqID := core.GenRequestID()
+
+	ch := make(chan *tractResult)
+	errc := make(chan core.Error)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	err := core.ErrAllocHost // default error if none present
+	for i, n := range order {
+		go func(i, j int) {
+			select {
+			case <-ctx.Done():
+				errc <- core.ErrCanceled
+				return
+			case <-cli.backupDelay(time.Duration(2*i) * time.Millisecond):
+				// sleep for the backup delay to prevent tied backup requests cancelling,
+				// each other's work.
+			}
+
+			host := tract.Hosts[j]
+			if host == "" {
+				log.V(1).Infof("read %s from tsid %d: no host", tract.Tract, tract.TSIDs[j])
+				return
+			}
+
+			// Compute alternate hosts that will receive backup requests. Pass them along
+			// to the tracserver rpc layer to enable cancellation messages to be sent between
+			// servers. TODO(eric): come up with a better name for this.
+			otherHosts := cutSlice(tract.Hosts, j)
+
+			_, badVersionHost, err = cli.readIntoReportBadTS(ctx, curAddr, host, otherHosts, reqID, result, tract, thisB, thisOffset)
```

Comment:
You're using the same `result` for all these calls, so there's a race there. You can fix that easily by using separate `result`s. But there's a worse problem: you're using the same buffer for all the calls to read into. Of course that's deliberate, since we're using ReadInto exactly to minimize copies. But you can't guarantee that another reply won't come back and get read into the same buffer before the cancel takes effect.

Unfortunately I don't see an easy way to fix this without some deep surgery (we could possibly do it in gobBulkCodec but it feels like it might get messy). So maybe for now we can accept the extra copy when we're doing backup requests, and later think about how to avoid it.

Reply:
Okay. That is a bit tricky. I will go with what you mention about accepting extra copies and not using the same buffer.